[MRG-1] ENH: Allow float32 to pass through with copy

jseabold opened the pull request:
This is the easy path to fixing this.

I was a bit surprised to see the float64 default and the copies elsewhere in the sparse functions. Are there rules for consistency as to float32 vs. float64 in the Cython code?

jseabold committed:
ENH: Allow float32 to pass through with copy

GaelVaroquaux commented:
In general the good looks good, but I would like to know the exact problem that you are trying to solve. Ideally, this problem should be illustrated by a test that fails without your patch and passes with it. The idea being that it will guide later refactoring.

In the long run, we want to tackle the issue of multiple types in Cython using fused types.

jseabold committed:
TST: Regression test for 32-bit input

jseabold committed:
DOC: Clarify doc for copy behavior

jseabold commented:
I found that computing cosine distance fails on 32-bit floating point sparse arrays. Added regression tests.

I added some clarifying documentation on the `assign_rows_csr` but perhaps this should fail given it's supposed to be no-copy?

Fused types would be nice.

MechCoder commented:
This fix should work, but I'm afraid that this is equivalent to changing the dtype of the sparse matrix prior and passing it to the function as it produces a copy.

It would be great if you could add fused type support so that we can do this without copying !

springcoil commented:
I had a quick look at this today - the code looks fine to me, and the fix works. I would recommend merging it anyway - but fused types needs done too.

ogrisel commented on sklearn/utils/sparsefuncs_fast.pyx:
```diff
@@ -378,7 +382,9 @@ def assign_rows_csr(X,
-        np.ndarray[DOUBLE, ndim=1] data = X.data
+        # might copy
+        np.ndarray[DOUBLE, ndim=1] data = np.asarray(X.data,
+                                                     dtype=np.float64)
```
This change to `assign_rows_csr` is not tested by the new regression test.

ogrisel commented on sklearn/utils/tests/test_sparsefuncs.py:
```diff
@@ -478,3 +480,10 @@ def test_csc_row_median():
+    # regression tests for passing 32-bit floating point
+    X = sp.rand(10, 5, dtype=np.float32, density=.5).tocsr()
+    inplace_csr_row_normalize_l1(X)
+    inplace_csr_row_normalize_l2(X)
```
This does not check that X has been properly normalized, and in practice it cannot be the case as the data is array which is actually normalized is a copy of the input `X.data` when the dtype is not `np.float64`.

ogrisel commented:
`inplace_csr_row_normalize_l1` and `inplace_csr_row_normalize_l2` are supposed to be inplace operations, therefore they cannot by design make a copy of the input data.

The only correct fix for those 2 functions is to use cython fused types for `np.float32` and `np.float64` (or directly `cython.floating` that should work for this case AFAIK).

Therefore I am -1 on this part of the PR.

+0 for the change to `assign_rows_csr` if a test is added, but I think it would be worth using fused types directly for this case as well.

GaelVaroquaux commented:
> inplace_csr_row_normalize_l1 and inplace_csr_row_normalize_l2 are
> supposed to be inplace operations, therefore they cannot by design make
> a copy of the input data.

> Therefore I am -1 on this part of the PR.

Agreed with the analysis. Sorry for my previous :+1:, it would have been
an error to have "in-place" functions copy data.

> The only correct fix for those 2 functions is to use cython fused types for
> np.float32 and np.float64 (or directly cython.floating that should work for
> this case AFAIK).

That would be good.

jnothman commented:
Can we close this as the incorrect patch, while waiting for a fix as part of @yenchenlin1994's GSoC?

jnothman commented:
That's what I'm doing anyway...
